database/merkledb/tree_persistence.go

```diff
@@ -0,0 +1,141 @@
+package merkledb
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/database/versiondb"
+
+	"github.com/ava-labs/avalanchego/codec/linearcodec"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ava-labs/avalanchego/database"
+)
+
+// TreePersistence holds the DB + the RootNode
+type TreePersistence struct {
+	db    database.Database
+	codec codec.Manager
+}
+
+// NewTreePersistence creates a new TreePersistence
+func NewTreePersistence(db database.Database) (Persistence, error) {
+	c := linearcodec.NewDefault()
+	_ = c.RegisterType(&BranchNode{})
+	_ = c.RegisterType(&LeafNode{})
+	_ = c.RegisterType(&RootNode{})
+	codecManager := codec.NewDefaultManager()
+	if err := codecManager.RegisterCodec(0, c); err != nil {
+		return nil, err
+	}
+
+	TreePersistence := TreePersistence{
+		db:    versiondb.New(db),
+		codec: codecManager,
+	}
+
+	return &TreePersistence, nil
+}
+
+// GetNodeByUnitKey fetches a Node given a StorageKey
+func (tp *TreePersistence) GetNodeByHash(nodeHash []byte) (Node, error) {
+	nodeBytes, err := tp.db.Get(nodeHash)
+	if err != nil {
+		return nil, err
+	}
+
+	var node Node
+	_, err = tp.codec.Unmarshal(nodeBytes, &node)
+	node.SetPersistence(tp)
```

Comment:
If `Unmarshal` returns an error, it is not safe to call `node.SetPersistence(fp)`.